pkg/service/backtest_sql.go

```diff
@@ -18,11 +18,24 @@ import (
 	"github.com/c9s/bbgo/pkg/types"
 )
 
+type IBacktestService interface {
+	Verify(sourceExchange types.Exchange, symbols []string, startTime time.Time, endTime time.Time) error
+	Sync(ctx context.Context, ex types.Exchange, symbol string, interval types.Interval, since, until time.Time) error
+	QueryKLine(ex types.ExchangeName, symbol string, interval types.Interval, orderBy string, limit int) (*types.KLine, error)
+	QueryKLinesForward(exchange types.ExchangeName, symbol string, interval types.Interval, startTime time.Time, limit int) ([]types.KLine, error)
+	QueryKLinesBackward(exchange types.ExchangeName, symbol string, interval types.Interval, endTime time.Time, limit int) ([]types.KLine, error)
+	QueryKLinesCh(since, until time.Time, exchange types.Exchange, symbols []string, intervals []types.Interval) (chan types.KLine, chan error)
+}
+
 type BacktestService struct {
 	DB *sqlx.DB
 }
 
-func (s *BacktestService) SyncKLineByInterval(ctx context.Context, exchange types.Exchange, symbol string, interval types.Interval, startTime, endTime time.Time) error {
+func NewBacktestService(db *sqlx.DB) IBacktestService {
```

Comment:
avoid to set interface type as the return type in the constructor, you can force the interface type from the caller. e.g.,

var s IBacktestService = NewBacktestService(db)

Reply:
and where would you take db from?

Reply:
it's initialized from the bbgo.Environment